File: src/main/java/org/sonar/plugins/xquery/checks/StrongTypingInModuleVariableCheck.java
/*
 * © 2013 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks;

import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.plugins.xquery.parser.XQueryParser;
import org.sonar.plugins.xquery.parser.XQueryTree;

/**
 * Checks for strong typing in module variable declarations
 * 
 * @since 1.0
 */
@Rule(
    key = "StrongTypingInModuleVariables",
    name = "Use Strong Typing when Declaring Module Variable",
    description = "Declare types for declared variables to increase readability and catch potential bugs. " +
            "Also try to scope the types as narrowly as possible " +
            "(i.e. use 'element()' instead of 'item()' when the value is an element) " +
            "and include quantifiers on each type.",
    priority = Priority.CRITICAL
)
public class StrongTypingInModuleVariableCheck extends AbstractCheck {

    @Override
    public void enterExpression(XQueryTree node) {
        super.enterExpression(node);
        
        // Process variable declarations
        if (XQueryParser.VarDecl == node.getType()) {
            
            // If we don't have the variable type then create a violation
            XQueryTree type = node.find("VarType");
            if (type == null || type.getChildCount() == 0) {
                createViolation(node.getLine());
            }
        }
    }
}

File: src/main/java/org/sonar/plugins/xquery/checks/StrongTypingInModuleVariableCheck.kt
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks

import org.antlr.v4.runtime.ParserRuleContext
import org.sonar.api.rule.RuleKey
import org.sonar.check.Rule
import org.sonar.plugins.xquery.checks.StrongTypingInModuleVariableCheck.Companion.RULE_KEY
import org.sonar.plugins.xquery.parser.XQueryParser
import org.sonar.plugins.xquery.parser.getLine
import org.sonar.plugins.xquery.rules.CheckClasses

/**
 * Checks for strong typing in module variable declarations
 *
 * @since 1.0
 */
@Rule(key = RULE_KEY)
class StrongTypingInModuleVariableCheck : AbstractCheck() {

    override fun enterExpression(node: ParserRuleContext) {
        super.enterExpression(node)

        // Process variable declarations
        if (node is XQueryParser.VarDeclContext) {

            // If we don't have the variable type then create a violation
            if (node.type == null) {
                createIssue(RULE, node.getLine())
            }
        }
    }

    companion object {
        const val RULE_KEY = "StrongTypingInModuleVariables"
        private val RULE = RuleKey.of(CheckClasses.REPOSITORY_KEY, RULE_KEY)
    }
}
